hack/tools/k8smaintainer/main.go

```diff
@@ -0,0 +1,324 @@
+			continue
+		}
+
+		if determinedVer == "" {
+			log.Printf("WARNING: Neither target version %s nor its predecessor found for %s. Skipping pinning.", targetStagingVer, effectivePath)
+			continue
+		}
+
+		if determinedVer != targetStagingVer {
+			log.Printf("WARNING: Target version %s not found for %s. Using existing version %s.", targetStagingVer, effectivePath, determinedVer)
+		}
+
+		// map the original module path (as seen in the dependency graph) to the desired version for the 'replace' directive
+		pins[mod.Path] = determinedVer
+	}
+
+	// Add k8s.io/kubernetes itself to the pins map
+	pins[k8sRepo] = k8sVer
+	log.Printf("Identified %d k8s.io/* modules to manage.", len(pins))
+
+	// 7. Parse go.mod again (to have a fresh modfile object)
+	modBytes, err = os.ReadFile("go.mod")
+	if err != nil {
+		log.Fatalf("Error reading go.mod again: %v", err)
+	}
+	modF, err = modfile.Parse("go.mod", modBytes, nil)
+	if err != nil {
+		log.Fatalf("Error parsing go.mod again: %v", err)
+	}
+
+	// Remove all existing k8s.io/* replaces
+	log.Println("Removing existing k8s.io/* replace directives...")
+	var replacesToRemove []string
+	for _, rep := range modF.Replace {
+		if strings.HasPrefix(rep.Old.Path, "k8s.io/") {
+			replacesToRemove = append(replacesToRemove, rep.Old.Path)
+		}
+	}
+	for _, path := range replacesToRemove {
+		if err := modF.DropReplace(path, ""); err != nil {
+			// Tolerate errors if the replace was already somehow removed
+			log.Printf("Note: Error dropping replace for %s (might be benign): %v", path, err)
+		}
+	}
+
+	// Add new replace directives
+	log.Println("Adding determined replace directives...")
+	// Sort for deterministic output
+	sortedPaths := make([]string, 0, len(pins))
+	for path := range pins {
+		sortedPaths = append(sortedPaths, path)
+	}
+	sort.Strings(sortedPaths)
+
+	for _, path := range sortedPaths {
+		version := pins[path]
+		// Add replace for the module path itself (e.g., k8s.io/api => k8s.io/api v0.32.3)
+		// This handles cases where the effective path from `go list` might differ due to other replaces
+		if err := modF.AddReplace(path, "", path, version); err != nil {
+			log.Fatalf("Error adding replace for %s => %s %s: %v", path, path, version, err)
+		}
+		log.Printf("Adding replace: %s => %s %s", path, path, version)
+	}
+
+	// Write go.mod
+	log.Println("Writing updated go.mod...")
+	modF.Cleanup() // Sort blocks, etc.
+	newModBytes, err := modF.Format()
+	if err != nil {
+		log.Fatalf("Error formatting go.mod: %v", err)
+	}
+	if err := os.WriteFile("go.mod", newModBytes, goModFilePerms); err != nil {
+		log.Fatalf("Error writing go.mod: %v", err)
+	}
+
+	// Run `go mod tidy`
+	goVer := modF.Go.Version
+	tidyArgs := []string{"mod", "tidy"}
+	if goVer != "" {
+		tidyArgs = append(tidyArgs, fmt.Sprintf("-go=%s", goVer))
+		log.Printf("Running 'go mod tidy -go=%s'...", goVer)
+	} else {
+		log.Println("Running 'go mod tidy'...")
+	}
```

Comment:
This could be done as a single `Printf()` using `join` on the strings array,

Reply:
fixing in next commit: 3a7c448b